README
This is a "plugin" for the Video Disk Recorder (VDR).

Written by:                  Luca De Pieri <dpluca@libero.it>

Project's homepage:          http://lukkinosat.altervista.org

Latest version available at: http://lukkinosat.altervista.org

See the file COPYING for license information.

Thanks to all users of community "VDR Italy" for tests - http://vdr-italia.phpbb24.com
Thanks to all users of irc channel for tests - #vdr irc.azzurra.org
Thanks to "Hans Dingemans" and "Luca Olivetti" for code and tests


Description:
============
This plugin load EPG data sent to MediaHighWay (Canal+ group)
or SkyBox (Sky - News Corporation) receivers,
or executes external scripts and load EPG data from file.


Notes:
======
This plugin work with vdr >= 1.4.x (tested with 1.4.6, 1.5.14, 1.6.0, 1.7.0)


Installation:
=============
- Unpack the archive to your plugins source directory.
- If this is first installation, the plugin create directory 'loadepg'
  to your plugins-directory (default '/video/plugins/loadepg') and
  copy configuration files, else is necessary to copy manually configuration
  files to your plugins-directory.
- Compile vdr and plugins as usual.


Commands OSD:
=============
Press Menu to select LoadEPG.
Press Up or Down arrow for selected a source.
Press Ok to start loading.
To stop loading, press Menu to select LoadEPG, and press Red.


Commands SVDRP:
===============
LIST - For list of sources available.
  example: svdrpsend.pl PLUG LOADEPG LIST
    
LOAD <SOURCE_ID> - For loading epg data from source selected.
  example: svdrpsend.pl PLUG LOADEPG LOAD 1


Configuration files:
====================
The configuration files are located into your plugins-directory (default '/video/plugins/loadepg').
Is possible specify an other directory for configuration files with the option '-c':
  example: ./vdr -P'loadepg -c /etc/vdr/plugins' ...

loadepg.conf - This file contains the list of sources for loading of epg data.
               For loading epg data sent to Mediahighway (Ver.1) receivers:
               MHW_1:<YourDescription>:<Frequency>:<Polarization>:<Source>:<SymbolRate>
    
               For loading epg data sent to Mediahighway (Ver.2) receivers:
               MHW_2:<YourDescription>:<Frequency>:<Polarization>:<Source>:<SymbolRate>

               For loading epg data sent to SkyBox receivers:
               SKYBOX:<YourDescription>:<Frequency>:<Polarization>:<Source>:<SymbolRate>:<DictionaryHuffamn>:<Themes>

loadepg.equiv - This file allows you to copy the epg data than one channel to
                another channel (equivalent channel). For example, the channel
                "Eurosport" is this in many bouquets, and you can make available
                the epg data also in other channels.
                Example:
                  S13.0E-64511-6700-11504 S13.0E-318-200-13831 Eurosport
                  
		  S13.0E-64511-6700-11504 is Channel ID of "Eurosport" on 
                  bouquet SkyItalia, and this is also copied on bouquet 
                  "EUTELSAT-RMB" with Channel ID S13.0E-318-200-13831

sky_it.dict - This file is dictionary Huffman for decoding epg data sent to
              SkyBox receivers. I am able to decode only for bouquet
              SkyItalia.

sky_XX.dict - Any other file dictionary Huffman for decoding epg data sent to
              SkyBox receivers in other countries. XX = language.

sky_it.themes - This file is table themes for bouquet SkyItalia.

sky_XX.themes - This file is table themes for other bouquets.


Debug:
======
After to loaded are available in /tmp 3 files:
epg.tmp - This file contains the epg data before of to be load in VDR.
epg.err - This file contains the errors of decoding Huffman code (only SkyBox).
epg.channels - This file contains the list channels available in epg data.



HISTORY
VDR Plugin 'loadepg' Revision History
-----------------------------------

2005-02-05: Version 0.1.0
- Initial revision.

2005-02-07: Version 0.1.1
- Correct time event.
- Fixed progress bar.
- Correct read configuration file with vdr 1.3.17.
- Added "Provider name" in progress bar.

2005-02-14: Version 0.1.2
- Fixed a bug when epg pid ( 210, 211 ) are egual to channel pid ( function cFilter::AttachFilter ).
- Added translation file ( file i18n.c ) for languages.

2005-03-30: Version 0.1.3
- Added setup menu.
- Added file loadepg.equiv for equivalences between two or more channels ( see loadepg.equiv file for info ).
- Added support for scanning on other dvb devices.

2005-05-10: Version 0.1.4
- Fixed error compilation with VDR += 1.3.24

2005-09-13: Version 0.1.5
- Fixed error compilation with VDR += 1.3.32

2005-10-05: Version 0.1.6
- Improved the source code for better preformances
- Added Finnish translation (by Rolf Ahrenberg)
- Added provider Canal Digital on S1W satellite (by Rolf Ahrenberg)

2006-02-09: Version 0.1.7
- Added support SVDRP, command LOAD <PROVIDER_NUMBER>
- Added German translation (by Ronny Kornexl)
- Update Finnish translation (by Rolf Ahrenberg)
- Remove Canal Digital on S1W satellite (my previous error :-) )

2006-04-26: Version 0.1.8
- Added option '-c=PATH' or '--config=PATH' to specify directory of configuration data
- Added timeout for no data epg available
- Update loadepg.conf for provider Sky Italia
- Added support for APIVERSION, VDR >= 1.3.47 (by Ronny Kornexl)

2006-06-15: Version 0.1.9
- Added support for data epg of Digital+ (S19.2E Spain - new mhwepg2)
- Increase the timeout to 30 seconds (by Dick Streefland)
- Added Dutch translation (by Dick Streefland)
- Added Spanish translation (by Fernando)
- Rewrite function set data transponder
- Rewrite function get time transponder

2007-05-31: Version 0.1.10
- Added support to execute external script and to load epg data file (option SCRIPT, i.e loadepg.conf)
- Added support to load epg data from external files (option FILE, i.e loadepg.conf)

2007-07-26: Version 0.1.11
- Fixed for new format epg of Digital+ (S19.2E Spain)

2008-02-08: Version 0.1.12
- Added support for APIVERSION, VDR >= 1.5.x

2008-06-16: Version 0.2.0
- Added support for data epg sent to SkyBox receivers (necessary dictionary huffman, at moment only for SkyItalia).
- Changed system OSD. Now is used the current skin.
- Added automatic scanning for search of a device available (see setup plugin).
- Added support for tuning with rotor (see setup plugin).
- Added support per APIVERSION, VDR >= 1.6.x and VDR >= 1.7.x

2008-09-15: Version 0.2.1pre01
- Added support for data epg sent to SkyBox UK

2009-07-31: Version 0.2.2pre01
- Fixed support for card DVB-S2

2010-03-01: Version 0.2.3
- Fixed function set data transponder for VDR >= 1.7.13


loadepg.hpp
//#include <vdr/plugin.h>
//#include <vdr/device.h>
//#include <vdr/remote.h>
//#include <vdr/thread.h>
#include <getopt.h>
#include <sys/ioctl.h>
#include <linux/dvb/frontend.h>
#include <linux/dvb/dmx.h>
#include <poll.h>
//#include "../../../libsi/section.h"

#include "formats.h"
//#include "i18n.h"

//#ifndef APIVERSNUM
//#include <vdr/config.h>
//#endif

#define DEBUG true
#define DEBUG_STARTTIME false

#define DVB_DEVICE_DEMUX "/dev/dvb/adapter%i/demux0"
#define LOADEPG_FILE_CONF "loadepg.conf"
#define LOADEPG_FILE_EQUIV "loadepg.equiv"
#define FILE_EPG_TMP "/tmp/epg.tmp"
#define FILE_EPG_ERR "/tmp/epg.err"
#define FILE_EPG_CHANNELS "/tmp/epg.channels"

#define MAX_FILTERS 24
#define MAX_ACTIVE_FILTERS 6
#define MAX_PROVIDERS 64

#define MAX_BOUQUETS 256
#define MAX_THEMES 4096
#define MAX_CHANNELS 4096
#define MAX_TITLES 262144
#define MAX_SUMMARIES 262144

#define MAX_BUFFER_SIZE_CHANNELS 1048576
#define MAX_BUFFER_SIZE_TITLES 4194304
#define MAX_BUFFER_SIZE_SUMMARIES 33554432

#define DATA_FORMAT_SKYBOX 1
#define DATA_FORMAT_MHW_1 2 
#define DATA_FORMAT_MHW_2 3
#define DATA_FORMAT_FILE 4
#define DATA_FORMAT_SCRIPT 5

#define TIMEOUT_CONTROL 90 // seconds
#define TIMEOUT_FILTER 5000 // ms
#define TIMEOUT_ROTOR 30 // seconds

struct sNode
{
  char *Value;
  struct sNode *P0;
  struct sNode *P1;
};

typedef struct sNode sNodeH;

typedef struct
{
  char *Directory;
  int DvbAdapterHasRotor;
  int DvbAdapterNumber;
  int EnableOsdMessages;
  int UseFileEquivalents;
} sConfig;

typedef struct
{
  unsigned short int OriginalSourceId;
  unsigned short int OriginalNid;
  unsigned short int OriginalTid;
  unsigned short int OriginalSid;
  unsigned short int OriginalRid;
  unsigned short int EquivSourceId;
  unsigned short int EquivNid;
  unsigned short int EquivTid;
  unsigned short int EquivSid;
  unsigned short int EquivRid;
} sEquivChannel;

typedef struct
{
  unsigned char Name[256];
} sTheme;

typedef struct
{
  unsigned short int ChannelId;
  unsigned short int Nid;
  unsigned short int Tid;
  unsigned short int Sid;
  unsigned short int SkyNumber;
  unsigned int pData;
  unsigned int lenData;
  bool IsFound;
  bool IsEpg;
} sChannel;

typedef struct
{
  unsigned short int BouquetId;
  short int SectionNumber[256];
  short int LastSectionNumber;
} sBouquet;

typedef struct
{
  unsigned short int ChannelId;
  unsigned short int ThemeId;
  unsigned short int MjdTime;
  unsigned short int EventId;
  unsigned int StartTime;
  unsigned int Duration;
  unsigned char SummaryAvailable;
  unsigned int pData;
  unsigned int lenData;
} sTitle;

typedef struct
{
  unsigned short int ChannelId;
  unsigned short int MjdTime;
  unsigned short int EventId;
  unsigned int pData;
  unsigned lenData;
} sSummary;

typedef struct
{
  char *Title;
  int DataFormat;
  int SourceId;
  char *Parm1;
  char *Parm2;
  char *Parm3;
} sProvider;

typedef struct
{
  int Fd;
  int Step;
  unsigned short int Pid;
  unsigned char Tid;
  unsigned char Mask;
} sFilter;

typedef struct
{
  int Fd;
  int FilterId;
  bool IsBusy;
} sActiveFilter;

class cControlLoadepg
{
  private:
  protected:
    virtual void Action( void );
  public:
    cControlLoadepg( void );
    ~cControlLoadepg();
    bool Loading;
    bool Stopping;
    bool Error;
};

class cTaskLoadepg
{
  private:
    int SetupUpdateChannels;
    int DvbAdapterNumber;
//    cChannel *VdrChannel;
//    cChannel *EpgChannel;
//    cDevice *EpgDevice;
    sNodeH H;
    sNodeH *nH;
    bool HasSwitched;
    bool IsError;
    bool IsRunning;
    bool EndBAT;
    bool EndThemes;
    bool EndChannels;
    int pC;
    int pT;
    int pS;
    struct pollfd PFD[MAX_FILTERS];
    int nFilters;
    sFilter Filters[MAX_FILTERS];
    int nActiveFilters;
    sActiveFilter ActiveFilters[MAX_ACTIVE_FILTERS];
    unsigned char InitialBuffer[MAX_FILTERS][64];
    void AddFilter( unsigned short int Pid, unsigned char Tid = 0x00, unsigned char Mask = 0xff );
    void StartFilter( int FilterId );
    void StopFilter( int ActiveFilterId );
    void PollingFilters( int Timeout );
    void ReadBuffer( int FilterId, int Fd );
    void GetLocalTimeOffset( void );
    void GetSatelliteTimeOffset( int FilterId, unsigned char *Data, int Length );
    void GetChannelsSKYBOX( int FilterId, unsigned char *Data, int Length );
    void GetTitlesSKYBOX( int FilterId, unsigned char *Data, int Length );
    void GetSummariesSKYBOX( int FilterId, unsigned char *Data, int Length );
    void GetThemesMHW1( int FilterId, unsigned char *Data, int Length );
    void GetChannelsMHW1( int FilterId, unsigned char *Data, int Length );
    void GetTitlesMHW1( int FilterId, unsigned char *Data, int Length );
    void GetSummariesMHW1( int FilterId, unsigned char *Data, int Length );
    void GetThemesMHW2( int FilterId, unsigned char *Data, int Length );
    void GetChannelsMHW2( int FilterId, unsigned char *Data, int Length );
    void GetTitlesMHW2( int FilterId, unsigned char *Data, int Length );
    void GetSummariesMHW2( int FilterId, unsigned char *Data, int Length );
    void CreateEpgDataFile( void );
    bool ReadFileDictionary( void );
    bool ReadFileThemes( void );
    void CleanString( unsigned char *String );
    unsigned char DecodeText[4096];
    unsigned char DecodeErrorText[4096];
    int DecodeHuffmanCode( unsigned char *Data, int Length );
    void CreateFileChannels( const char *FileChannels );
  protected:
    virtual void Action( void );
  public:
    cTaskLoadepg( void );
    ~cTaskLoadepg();
    void LoadFromSatellite( void );
    void LoadFromFile( const char *FileEpg );
    void LoadFromScript( const char *FileScript, const char *FileEpg );
};

class cMainMenuLoadepg
{
  private:
  protected:
  public:
    cMainMenuLoadepg( void );
    ~cMainMenuLoadepg();
//    eOSState ProcessKey( eKeys Key );
};

class cSetupMenuLoadepg 
{
  private:
    const char *dvbAdapterNumberFormatTexts[8];
    const char *dvbAdapterHasRotorFormatTexts[8];
  protected:
    virtual void Store( void );
  public:
    cSetupMenuLoadepg( void );
    ~cSetupMenuLoadepg();
};


loadepg.h
//#include <vdr/plugin.h>
//#include <vdr/device.h>
//#include <vdr/remote.h>
//#include <vdr/thread.h>
#include <getopt.h>
#include <sys/ioctl.h>
#include <linux/dvb/frontend.h>
#include <linux/dvb/dmx.h>
//#include "../../../libsi/section.h"

#include "formats.h"
#include "i18n.h"

//#ifndef APIVERSNUM
//#include <vdr/config.h>
//#endif

#define DEBUG true
#define DEBUG_STARTTIME false

#define DVB_DEVICE_DEMUX "/dev/dvb/adapter%i/demux0"
#define LOADEPG_FILE_CONF "loadepg.conf"
#define LOADEPG_FILE_EQUIV "loadepg.equiv"
#define FILE_EPG_TMP "/tmp/epg.tmp"
#define FILE_EPG_ERR "/tmp/epg.err"
#define FILE_EPG_CHANNELS "/tmp/epg.channels"

#define MAX_FILTERS 24
#define MAX_ACTIVE_FILTERS 6
#define MAX_PROVIDERS 64

#define MAX_BOUQUETS 256
#define MAX_THEMES 4096
#define MAX_CHANNELS 4096
#define MAX_TITLES 262144
#define MAX_SUMMARIES 262144

#define MAX_BUFFER_SIZE_CHANNELS 1048576
#define MAX_BUFFER_SIZE_TITLES 4194304
#define MAX_BUFFER_SIZE_SUMMARIES 33554432

#define DATA_FORMAT_SKYBOX 1
#define DATA_FORMAT_MHW_1 2 
#define DATA_FORMAT_MHW_2 3
#define DATA_FORMAT_FILE 4
#define DATA_FORMAT_SCRIPT 5

#define TIMEOUT_CONTROL 90 // seconds
#define TIMEOUT_FILTER 5000 // ms
#define TIMEOUT_ROTOR 30 // seconds

struct sNode
{
  char *Value;
  struct sNode *P0;
  struct sNode *P1;
};

typedef struct sNode sNodeH;

typedef struct
{
  char *Directory;
  int DvbAdapterHasRotor;
  int DvbAdapterNumber;
  int EnableOsdMessages;
  int UseFileEquivalents;
} sConfig;

typedef struct
{
  unsigned short int OriginalSourceId;
  unsigned short int OriginalNid;
  unsigned short int OriginalTid;
  unsigned short int OriginalSid;
  unsigned short int OriginalRid;
  unsigned short int EquivSourceId;
  unsigned short int EquivNid;
  unsigned short int EquivTid;
  unsigned short int EquivSid;
  unsigned short int EquivRid;
} sEquivChannel;

typedef struct
{
  unsigned char Name[256];
} sTheme;

typedef struct
{
  unsigned short int ChannelId;
  unsigned short int Nid;
  unsigned short int Tid;
  unsigned short int Sid;
  unsigned short int SkyNumber;
  unsigned int pData;
  unsigned int lenData;
  bool IsFound;
  bool IsEpg;
} sChannel;

typedef struct
{
  unsigned short int BouquetId;
  short int SectionNumber[256];
  short int LastSectionNumber;
} sBouquet;

typedef struct
{
  unsigned short int ChannelId;
  unsigned short int ThemeId;
  unsigned short int MjdTime;
  unsigned short int EventId;
  unsigned int StartTime;
  unsigned int Duration;
  unsigned char SummaryAvailable;
  unsigned int pData;
  unsigned int lenData;
} sTitle;

typedef struct
{
  unsigned short int ChannelId;
  unsigned short int MjdTime;
  unsigned short int EventId;
  unsigned int pData;
  unsigned lenData;
} sSummary;

typedef struct
{
  char *Title;
  int DataFormat;
  int SourceId;
  char *Parm1;
  char *Parm2;
  char *Parm3;
} sProvider;

typedef struct
{
  int Fd;
  int Step;
  unsigned short int Pid;
  unsigned char Tid;
  unsigned char Mask;
} sFilter;

typedef struct
{
  int Fd;
  int FilterId;
  bool IsBusy;
} sActiveFilter;

class cControlLoadepg :
{
  private:
  protected:
    virtual void Action( void );
  public:
    cControlLoadepg( void );
    ~cControlLoadepg();
    bool Loading;
    bool Stopping;
    bool Error;
};

class cTaskLoadepg :
{
  private:
    int SetupUpdateChannels;
    int DvbAdapterNumber;
    cChannel *VdrChannel;
    cChannel *EpgChannel;
    cDevice *EpgDevice;
    sNodeH H;
    sNodeH *nH;
    bool HasSwitched;
    bool IsError;
    bool IsRunning;
    bool EndBAT;
    bool EndThemes;
    bool EndChannels;
    int pC;
    int pT;
    int pS;
    struct pollfd PFD[MAX_FILTERS];
    int nFilters;
    sFilter Filters[MAX_FILTERS];
    int nActiveFilters;
    sActiveFilter ActiveFilters[MAX_ACTIVE_FILTERS];
    unsigned char InitialBuffer[MAX_FILTERS][64];
    void AddFilter( unsigned short int Pid, unsigned char Tid = 0x00, unsigned char Mask = 0xff );
    void StartFilter( int FilterId );
    void StopFilter( int ActiveFilterId );
    void PollingFilters( int Timeout );
    void ReadBuffer( int FilterId, int Fd );
    void GetLocalTimeOffset( void );
    void GetSatelliteTimeOffset( int FilterId, unsigned char *Data, int Length );
    void GetChannelsSKYBOX( int FilterId, unsigned char *Data, int Length );
    void GetTitlesSKYBOX( int FilterId, unsigned char *Data, int Length );
    void GetSummariesSKYBOX( int FilterId, unsigned char *Data, int Length );
    void GetThemesMHW1( int FilterId, unsigned char *Data, int Length );
    void GetChannelsMHW1( int FilterId, unsigned char *Data, int Length );
    void GetTitlesMHW1( int FilterId, unsigned char *Data, int Length );
    void GetSummariesMHW1( int FilterId, unsigned char *Data, int Length );
    void GetThemesMHW2( int FilterId, unsigned char *Data, int Length );
    void GetChannelsMHW2( int FilterId, unsigned char *Data, int Length );
    void GetTitlesMHW2( int FilterId, unsigned char *Data, int Length );
    void GetSummariesMHW2( int FilterId, unsigned char *Data, int Length );
    void CreateEpgDataFile( void );
    bool ReadFileDictionary( void );
    bool ReadFileThemes( void );
    void CleanString( unsigned char *String );
    unsigned char DecodeText[4096];
    unsigned char DecodeErrorText[4096];
    int DecodeHuffmanCode( unsigned char *Data, int Length );
    void CreateFileChannels( const char *FileChannels );
  protected:
    virtual void Action( void );
  public:
    cTaskLoadepg( void );
    ~cTaskLoadepg();
    void LoadFromSatellite( void );
    void LoadFromFile( const char *FileEpg );
    void LoadFromScript( const char *FileScript, const char *FileEpg );
};

class cMainMenuLoadepg : public cOsdMenu
{
  private:
  protected:
  public:
    cMainMenuLoadepg( void );
    ~cMainMenuLoadepg();
    eOSState ProcessKey( eKeys Key );
};

class cSetupMenuLoadepg : public cMenuSetupPage
{
  private:
    const char *dvbAdapterNumberFormatTexts[8];
    const char *dvbAdapterHasRotorFormatTexts[8];
  protected:
    virtual void Store( void );
  public:
    cSetupMenuLoadepg( void );
    ~cSetupMenuLoadepg();
};
